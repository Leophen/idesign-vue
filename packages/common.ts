import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
