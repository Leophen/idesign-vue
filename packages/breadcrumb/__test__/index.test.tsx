import { Breadcrumb, BreadcrumbItem } from '../index';
import { render } from '../../common'

describe('Breadcrumb ç»„ä»¶æµ‹è¯•', () => {
  const renderItem = (breadcrumb) => {
    const wrapper = render(breadcrumb);
    return wrapper.find('.i-breadcrumb__item')
  }

  it('create', () => {
    const wrapper = render(
      <Breadcrumb>
        <BreadcrumbItem>item1</BreadcrumbItem>
        <BreadcrumbItem>item2</BreadcrumbItem>
        <BreadcrumbItem>item3</BreadcrumbItem>
      </Breadcrumb>
    );
    expect(wrapper.element).toMatchSnapshot();
  });

  it('breadcrumbItem children', () => {
    const el = (
      <Breadcrumb>
        <BreadcrumbItem>item</BreadcrumbItem>
      </Breadcrumb>
    );
    expect(renderItem(el).text()).toBe('item');
  });

  it('breadcrumbItem disabled', () => {
    const el = (
      <Breadcrumb>
        <BreadcrumbItem disabled>item</BreadcrumbItem>
      </Breadcrumb>
    );
    expect(renderItem(el).classes()).toContain('i-breadcrumb-is-disabled');
  });

  it('max width', () => {
    const wrapper = render(
      <Breadcrumb maxItemWidth="120px">
        <BreadcrumbItem>item</BreadcrumbItem>
      </Breadcrumb>
    );
    expect(wrapper.find('.i-breadcrumb__inner').attributes('style')).toMatch('max-width: 120px;');
  });

  it('separator', () => {
    const wrapper = render(
      <Breadcrumb
        v-slots={{
          separator: () => <span>ğŸ‘‰</span>,
        }}
      >
        <BreadcrumbItem>item</BreadcrumbItem>
      </Breadcrumb>
    );
    expect(wrapper.find('.i-breadcrumb__separator').html()).toMatch('<span>ğŸ‘‰</span>');
  });
});

