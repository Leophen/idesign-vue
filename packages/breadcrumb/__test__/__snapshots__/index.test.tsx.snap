// Vitest Snapshot v1

exports[`Breadcrumb 组件测试 > create 1`] = `
<div
  class="i-breadcrumb"
>
  
  <div
    class="i-breadcrumb__item"
  >
    <span
      class="i-breadcrumb__inner"
      style="max-width: 200px;"
    >
      
      item1
      
    </span>
    <span
      class="i-breadcrumb__separator"
    >
      <i
        class="i-icon icon-ArrowRight"
        style="font-size: 16px;"
      />
    </span>
  </div>
  <div
    class="i-breadcrumb__item"
  >
    <span
      class="i-breadcrumb__inner"
      style="max-width: 200px;"
    >
      
      item2
      
    </span>
    <span
      class="i-breadcrumb__separator"
    >
      <i
        class="i-icon icon-ArrowRight"
        style="font-size: 16px;"
      />
    </span>
  </div>
  <div
    class="i-breadcrumb__item"
  >
    <span
      class="i-breadcrumb__inner"
      style="max-width: 200px;"
    >
      
      item3
      
    </span>
    <span
      class="i-breadcrumb__separator"
    >
      <i
        class="i-icon icon-ArrowRight"
        style="font-size: 16px;"
      />
    </span>
  </div>
  
</div>
`;
